### PR TITLE
Create VM from command line arguments

### PR DESCRIPTION
**Pull request by QiliangFan**
Create a command line parser to start a micro vm

Signed-off-by: qiliangfan <fanqiliang@mail.nankai.edu.cn>

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+nix = "0.24.1"
+anyhow = "1.0.61"
+vmm-sys-util = "0.9.0"
+slog = "2.7.0"
+slog-term = "2.9.0"
+slog-json = "2.6.1"
+slog-scope = "4.4.0"
+
+[patch.'crates-io']
+dbs-device = { git = "https://github.com/openanolis/dragonball-sandbox.git", rev = "7a8e832b53d66994d6a16f0513d69f540583dcd0" }
```
Comment (wllenyj) [inline Cargo.toml]: This `rev` needs the same as kata.
The current `c3d7831aee7c3962b8a90f0afbfd0fb7e4d30323` used by kata

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+# Copyright (c) 2019-2022 Ant Group. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+default: build
+
+build:
+	# FIXME: This line will be removed when we solve the vm-memory dependency problem in Dragonball Sandbox
+	cargo update -p vm-memory:0.8.0 --precise 0.7.0 || true
+	cargo update -p fuse-backend-rs --precise 0.9.4 || true
+	cargo update -p vmm-sys-util --precise 0.9.0 || true
```
Comment (wllenyj) [inline Makefile]: remove `cargo update`
Currently, this is no longer required to control the version.

Comment (wllenyj) [reply Makefile]: For generated binary projects, The `Cargo.lock` file can be submitted.
See [Why do binaries have Cargo.lock in version control, but not libraries?](https://doc.rust-lang.org/cargo/faq.html#why-do-binaries-have-cargolock-in-version-control-but-not-libraries) for details.

**Review by wllenyj — COMMENTED** [review]
Need to run `cargo clippy` checks.

**Comment by studychao** [comment]
Thanks ! @QiliangFan .
This is a really awesome tool for testing and debugging Dragonball !
But right now this could only be used upon dragonball in Kata repo's main branch,  is there a way to config this to connect to the Dragonball elsewhere ( e.g. in other fork from Kata repo or in my local environment)

**Comment by studychao** [comment]
Besides, you need to updete the commit id for dbs crates in your Cargo.toml

**Comment by QiliangFan** [comment]
> Thanks ! @QiliangFan . This is a really awesome tool for testing and debugging Dragonball ! But right now this could only be used upon dragonball in Kata repo's main branch, is there a way to config this to connect to the Dragonball elsewhere ( e.g. in other fork from Kata repo or in my local environment)

```rust
use dragonball::{
    api::v1::{
        BlockDeviceConfigInfo, BootSourceConfig, InstanceInfo, VmmAction, VmmActionError, VmmData,
        VmmRequest, VmmResponse,
    },
    vm::{CpuTopology, VmConfigInfo},
};
```

As shown in the above snippet, `dbs-cli` is a tool that passes serialized configuration objects to relied low-level crates. Any repos forked from [kata-containers](https://github.com/kata-containers/kata-containers/) should be able treated as alternatives (As there is a patch solving the problem about the connection of stdio stream and serial console). If needed, the sources of dependencies `dragonball` and `hypervisor`(required by `dragonball`) can be replaced with sources that meet the above requirement. 

If the demand is to change sources of dependence after compilation, I think it is a little complicated, and right now have no idea about it.

**Comment by studychao** [comment]
> > Thanks ! @QiliangFan . This is a really awesome tool for testing and debugging Dragonball ! But right now this could only be used upon dragonball in Kata repo's main branch, is there a way to config this to connect to the Dragonball elsewhere ( e.g. in other fork from Kata repo or in my local environment)
> 
> ```rust
> use dragonball::{
>     api::v1::{
>         BlockDeviceConfigInfo, BootSourceConfig, InstanceInfo, VmmAction, VmmActionError, VmmData,
>         VmmRequest, VmmResponse,
>     },
>     vm::{CpuTopology, VmConfigInfo},
> };
> ```
> 
> As shown in the above snippet, `dbs-cli` is a tool that passes serialized configuration objects to relied low-level crates. Any repos forked from [kata-containers](https://github.com/kata-containers/kata-containers/) should be able treated as alternatives (As there is a patch solving the problem about the connection of stdio stream and serial console). If needed, the sources of dependencies `dragonball` and `hypervisor`(required by `dragonball`) can be replaced with sources that meet the above requirement.
> 
> If the demand is to change sources of dependence after compilation, I think it is a little complicated, and right now have no idea about it.

Sure, then you need to write down in document to teach others how to use dbs-cli on their own `Dragonball`:)

**Comment by QiliangFan** [comment]
> > > Thanks ! @QiliangFan . This is a really awesome tool for testing and debugging Dragonball ! But right now this could only be used upon dragonball in Kata repo's main branch, is there a way to config this to connect to the Dragonball elsewhere ( e.g. in other fork from Kata repo or in my local environment)
> > 
> > 
> > ```rust
> > use dragonball::{
> >     api::v1::{
> >         BlockDeviceConfigInfo, BootSourceConfig, InstanceInfo, VmmAction, VmmActionError, VmmData,
> >         VmmRequest, VmmResponse,
> >     },
> >     vm::{CpuTopology, VmConfigInfo},
> > };
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > As shown in the above snippet, `dbs-cli` is a tool that passes serialized configuration objects to relied low-level crates. Any repos forked from [kata-containers](https://github.com/kata-containers/kata-containers/) should be able treated as alternatives (As there is a patch solving the problem about the connection of stdio stream and serial console). If needed, the sources of dependencies `dragonball` and `hypervisor`(required by `dragonball`) can be replaced with sources that meet the above requirement.
> > If the demand is to change sources of dependence after compilation, I think it is a little complicated, and right now have no idea about it.
> 
> Sure, then you need to write down in document to teach others how to use dbs-cli on their own `Dragonball`:)

Thanks, I will add it to document.

**Comment by studychao** [comment]
You could do this in another PR. This one is already big enough

**Comment by studychao** [comment]
issue #2 created

File: src/cli_instance.rs [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+                dies_per_socket: args.create_args.cpu_topology.dies_per_socket,
+                sockets: args.create_args.cpu_topology.sockets,
+            },
+            vpmu_feature: 0,
+            mem_type: args.create_args.mem_type.clone(),
+            mem_file_path: args.create_args.mem_file_path.clone(),
+            mem_size_mib: args.create_args.mem_size,
+            // as in crate `dragonball` serial_path will be assigned with a default value,
+            // we need a special token to enable the stdio console.
+            serial_path: Some("stdio".to_owned()),
```
Comment (studychao) [inline src/cli_instance.rs]: There should be some choices for user if they want to start a Dragonball VM but use vsock to connect to it

Comment (QiliangFan) [reply src/cli_instance.rs]: :ok:

Comment (QiliangFan) [reply src/cli_instance.rs]:
```rust
            // as in crate `dragonball` serial_path will be assigned with a default value,
             // we need a special token to enable the stdio console.
             serial_path: Some(args.create_args.serial_path.clone()),
         };

         // check the existence of the serial path (rm it if exist)
         let serial_file = Path::new(&args.create_args.serial_path);
         if args.create_args.serial_path != *"stdio" && serial_file.exists() {
             std::fs::remove_file(serial_file).unwrap();
         }
```

Comment (QiliangFan) [reply src/cli_instance.rs]: In `args.rs`:

```rust

   // The serial path used to communicate with VM
   #[clap(
       short,
       long,
       value_parser,
       default_value = "stdio",
       help = "The serial path used to communicate with VM",
       display_order = 2
   )]
   pub serial_path: String,
```

Comment (studychao) [reply src/cli_instance.rs]: please update the document

**Comment by studychao** [comment]
There are also some VMMAction from Dragonball like `UpdateNetworkInterface` that will update the VM after the VM boots up. Is there a way for dbs-cli to trigger those commands?

**Comment by QiliangFan** [comment]
> There are also some VMMAction from Dragonball like `UpdateNetworkInterface` that will update the VM after the VM boots up. Is there a way for dbs-cli to trigger those commands?

Documents have been updated.

It seems that we need a system-level service to achieve this goal. Otherwise, we need to open two terminal: the first serve as a server; the second act as a client to change the status of the VM (just like the firecracker). In this way, the current tool needs to change how it works. It can be implemented in this way, but why not just type a new command to start a vm again, as the rootfs retain the previous operations. (The tests for hot-plugging are more suitable for a client-server mode) 😢

**Comment by wllenyj** [comment]
Please rebase the latest main branch. It added ci support.

**Review by wllenyj — APPROVED** [review]

**Review by studychao — APPROVED** [review]
Thanks!! That's really helpful and hope we can do more work upon this!

**Comment by QiliangFan** [comment]
> Thanks!! That's really helpful and hope we can do more work upon this!

The extended work will be submitted gradually.

**Pull request merged**